drivers/watchdog/sunxi-twd.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright © 2017 Samuel Holland <samuel@sholland.org>
+ * Copyright © 2018 Drew Walters <drewwalters96@gmail.com>
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <compiler.h>
+#include <error.h>
+#include <mmio.h>
+#include <drivers/clock.h>
+#include <drivers/watchdog.h>
+#include <drivers/watchdog/sunxi-twd.h>
+
+#define R_TWD_CTRL_REG      (0x10)
+#define R_TWD_RESTART_REG   (0x14)
+#define R_TWD_INTV_VAL_REG  (0x30)
+
+#define TWD_RESTART_KEY_VAL (0xD14 << 15) | BIT(0)
+
+static int
+sunxi_twd_disable(struct device *dev)
+{
+	uint32_t reg;
+
+	/* Stop trusted watchdog. */
+	reg  = mmio_read32(dev->regs + R_TWD_CTRL_REG);
+	reg |= BIT(1);
+
+	mmio_write32(dev->regs + R_TWD_CTRL_REG, reg);
+
+	return SUCCESS;
+}
+
+static int
+sunxi_twd_enable(struct device *dev, uint32_t timeout)
+{
+	uint32_t reg;
+	uint32_t timeout_intv;
+
+	/* Convert timeout value from microseconds to timeout interval.
+	 *
+	 * TODO(@smaeul): Use clock_get_freq(dev) once clock driver is
+	 *                written.
+	 */
+	timeout_intv = timeout * 24;
+	mmio_write32(dev->regs + R_TWD_INTV_VAL_REG, timeout_intv);
+
+	/* Reset enable. */
+	reg  = mmio_read32(dev->regs + R_TWD_CTRL_REG);
+	reg |= BIT(8);
+
+	/* Enable and perform restart. */
```

Comment:
This should be extracted to its own function and also put in watchdog_driver_ops because it needs to be called periodically.

Reply:
Done.